bundles/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java

```diff
@@ -157,12 +159,13 @@ public void requestFullState() {
         if (config.apikey == null) {
             return;
         }
-        String url = url(config.host, config.apikey, null, null);
-
+        String url = url(config.host, config.port, config.apikey, null, null);
         http.get(url, config.timeout).thenApply(this::parseBridgeFullStateResponse).exceptionally(e -> {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
-            if (!(e instanceof SocketTimeoutException)) {
+            if (!(e instanceof SocketTimeoutException) && !(e instanceof TimeoutException)
+                    && !(e instanceof CompletionException)) {
                 logger.warn("Get full state failed", e);
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```

Comment:
The idea is not to go offline on timeout exceptions?

Reply:
Yes, it is. I am wondering if it is possible to see a `SocketTimeoutException` on a HTTP request.

I do not think the timeout happened because of a missing API key - at least in my case. Thus I additionally removed the `requestApiKey()` later in the code.

I saw two warnings directly behind each other in my case:

```
2019-08-29 22:09:59.102 [WARN ] [internal.handler.DeconzBridgeHandler] - Get full state failed
java.util.concurrent.CompletionException: java.util.concurrent.TimeoutException: Total timeout 2000 ms elapsed
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) [?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) [?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient$1.onComplete(AsyncHttpClient.java:113) [297:org.openhab.binding.deconz:2.5.0.M2]
        at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:196) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:188) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:441) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpReceiver.abort(HttpReceiver.java:523) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpChannel.abortResponse(HttpChannel.java:156) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpChannel.abort(HttpChannel.java:149) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpExchange.abort(HttpExchange.java:257) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpConversation.abort(HttpConversation.java:149) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.HttpRequest.abort(HttpRequest.java:766) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.client.TimeoutCompleteListener.onTimeoutExpired(TimeoutCompleteListener.java:51) [76:org.eclipse.jetty.client:9.4.18.v20190429]
        at org.eclipse.jetty.io.CyclicTimeout$Wakeup.run(CyclicTimeout.java:282) [80:org.eclipse.jetty.io:9.4.18.v20190429]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.util.concurrent.TimeoutException: Total timeout 2000 ms elapsed
        ... 9 more
2019-08-29 22:10:09.276 [WARN ] [internal.handler.DeconzBridgeHandler] - Authorisation failed
java.util.concurrent.CompletionException: org.eclipse.jetty.io.EofException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:647) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient$1.onComplete(AsyncHttpClient.java:113) ~[?:?]
        at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:196) ~[?:?]
        at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:188) ~[?:?]
        at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:441) ~[?:?]
        at org.eclipse.jetty.client.HttpReceiver.abort(HttpReceiver.java:523) ~[?:?]
        at org.eclipse.jetty.client.HttpChannel.abortResponse(HttpChannel.java:156) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.terminateRequest(HttpSender.java:396) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.abort(HttpSender.java:581) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.lambda$executeAbort$0(HttpSender.java:360) ~[?:?]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698) [92:org.eclipse.jetty.util:9.4.18.v20190429]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804) [92:org.eclipse.jetty.util:9.4.18.v20190429]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: org.eclipse.jetty.io.EofException
        at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:284) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:393) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277) ~[?:?]
        at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:380) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:268) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:214) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:85) ~[?:?]
        at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:128) ~[?:?]
        at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:201) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:255) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:122) ~[?:?]
        at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:346) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:304) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:294) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:269) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:246) ~[?:?]
        at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:576) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:726) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:718) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.doNetwork(AsyncHttpClient.java:107) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.post(AsyncHttpClient.java:53) ~[?:?]
        at org.openhab.binding.deconz.internal.handler.DeconzBridgeHandler.requestApiKey(DeconzBridgeHandler.java:239) ~[?:?]
        at org.openhab.binding.deconz.internal.handler.DeconzBridgeHandler.lambda$0(DeconzBridgeHandler.java:119) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:?]
        ... 1 more
Caused by: java.nio.channels.ClosedByInterruptException
        at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:511) ~[?:?]
        at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:264) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:393) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277) ~[?:?]
        at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:380) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:268) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:214) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:85) ~[?:?]
        at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:128) ~[?:?]
        at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:201) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:255) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:122) ~[?:?]
        at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:346) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:304) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:294) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:269) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:246) ~[?:?]
        at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:576) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:726) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:718) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.doNetwork(AsyncHttpClient.java:107) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:214) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:85) ~[?:?]
        at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:128) ~[?:?]
        at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:201) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:255) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:122) ~[?:?]
        at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:346) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:304) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:294) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:269) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:246) ~[?:?]
        at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:576) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:726) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:718) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.doNetwork(AsyncHttpClient.java:107) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.post(AsyncHttpClient.java:53) ~[?:?]
        at org.openhab.binding.deconz.internal.handler.DeconzBridgeHandler.requestApiKey(DeconzBridgeHandler.java:239) ~[?:?]
        at org.openhab.binding.deconz.internal.handler.DeconzBridgeHandler.lambda$0(DeconzBridgeHandler.java:119) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:?]
        ... 1 more
Caused by: java.nio.channels.ClosedByInterruptException
        at java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:202) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:511) ~[?:?]
        at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:264) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:393) ~[?:?]
        at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277) ~[?:?]
        at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:380) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:268) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[?:?]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) ~[?:?]
        at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62) ~[?:?]
        at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:214) ~[?:?]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:85) ~[?:?]
        at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:128) ~[?:?]
        at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:201) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:255) ~[?:?]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:122) ~[?:?]
        at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:346) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:304) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:294) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:269) ~[?:?]
        at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:246) ~[?:?]
        at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:576) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:726) ~[?:?]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:718) ~[?:?]
        at org.openhab.binding.deconz.internal.netutils.AsyncHttpClient.doNetwork(AsyncHttpClient.java:107) ~[?:?]
```

Reply:
A missing api key might indeed not being related to a socket or http timeout.
I didn't really check the http response in this implementation. It is probably 401 or similar if the api key need to be re-requested.

But I do wonder why we get those spurious timeouts. If I use curl I can request the http endpoint pretty fast without a single timeout.

Reply:
I have the slight feeling that it is based on the same root cause like https://github.com/openhab/openhab2-addons/issues/4769. The deCONZ REST API seem to have unexpected but recurring down times ...

Reply:
Maybe related to this: https://dresden-elektronik.github.io/deconz-rest-doc/polling/

Reply:
Forget about my last comment.

Currently I have the slight feeling that the deCONZ REST interface cannot handle all the requests which are triggered during restart at the same time if you have a lot of devices and some of them may fail. If they do the connection to the websocket for the sensor will never be established for future updates. I changed that too. The sensor will schedule a new initialization of itself.